sys/include/net/asymcute.h

```diff
@@ -0,0 +1,568 @@
+ * @return  false if topic is not registered
+ */
+static inline bool asymcute_topic_is_reg(const asymcute_topic_t *topic)
+{
+    assert(topic);
+    return (topic->con != NULL);
+}
+
+/**
+ * @brief   Check if a given topic is initialized
+ *
+ * @param[in] topic     topic to check
+ *
+ * @return  true if topic is initialized
+ * @return  false if topic is not initialized
+ */
+static inline bool asymcute_topic_is_init(const asymcute_topic_t *topic)
+{
+    assert(topic);
+    return (strlen(topic->name) > 0);
+}
+
+/**
+ * @brief   Compare two given topics and check if they are equal
+ *
+ * @param[in] a     topic A
+ * @param[in] b     topic B
+ *
+ * @return  true if both topics are equal
+ * @return  false if topics differ
+ */
+static inline bool asymcute_topic_equal(const asymcute_topic_t *a,
+                                        const asymcute_topic_t *b)
+{
+    assert(a);
+    assert(b);
+
+    return ((a->flags == b->flags) && (a->id == b->id));
+}
+
+/**
+ * @brief   Initialize the given topic
+ *
+ * @param[out] topic        topic to initialize
+ * @param[in] topic_name    topic name (ASCII), may be NULL if topic should use
+ *                          a pre-shared topic ID
+ * @param[in] topic_id      pre-shared topic ID, or don't care if @p topic_name
+ *                          is given
+ */
+int asymcute_topic_init(asymcute_topic_t *topic, const char *topic_name,
+                        uint16_t topic_id);
+
+/**
+ * @brief   Start a listener thread
+ *
+ * @note    Must have higher priority then the handler thread
+ *
+ * @param[in] con       connection context to use for this connection
+ * @param[in] priority  priority of the listener thread created by this function
+ * @param[in] callback  user callback for notification about connection related
+ *                      events
+ *
+ * @return  ASYMCUTE_OK on success
+ * @return  ASYMCUTE_BUSY if connection context is already in use
+ */
+int asymcute_listener_run(asymcute_con_t *con, char priority,
+                          asymcute_evt_cb_t callback);
+
+/**
+ * @brief   Start the global Asymcute handler thread for processing timeouts and
+ *          keep alive events
+ *
+ * This function is typically called during system initialization.
+ *
+ * @note    Must have lower priority than listener threads
+ */
+void asymcute_handler_run(void);
+
+/**
+ * @brief   Check if the given connection context is connected to a gateway
+ *
+ * @param[in] con       connection to check
+ *
+ * @return  true if context is connected
+ * @return  false if not connected
+ */
+bool asymcute_is_connected(const asymcute_con_t *con);
+
+/**
+ * @brief   Connect to the given MQTT-SN gateway
+ *
+ * @param[in,out] con   connection to use
+ * @param[in,out] req   request context to use for CONNECT procedure
+ * @param[in] server    UDP endpoint of the target gateway
+ * @param[in] cli_id    client ID to register with the gateway
+ * @param[in] clean     set `true` to start a clean session
```

Comment:
see the spec

Reply:
Since this is a user facing API, I would at least expect a sentence or two on what this means. At least name the section of the spec. However, requiring the user to read the spec can already be IMO booked under requiring "expert knowledge" (which a user API shouldn't expect).

Reply:
I really disagree here: our API doc is not supposed to explain how a certain protocol etc works, but it should give users that know the protocol the needed infos to apply it. I also do not want to explain what the different QoS levels mean - that is something the spec is for. Same is true for the clean session flag!

Reply:
Fine… I still disagree—an API should enable the user to use a component without expert knowledge—but then lets agree to disagree.

Reply:
:-)